File: 71.simplify-path.py
#
# @lc app=leetcode id=71 lang=python3
#
# [71] Simplify Path
#

# Time:  O(n)
# Space: O(n)

# @lc code=start
class Solution:
    def simplifyPath(self, path: str) -> str:
        stack, tokens = [], path.split("/")
        for token in tokens:
            if token == "..":
                if stack:
                    stack.pop()
            elif token and token != ".":
                stack.append(token)
        return "/" + "/".join(stack)


# @lc code=end

File: 71.simplify-path.js
/*
 * @lc app=leetcode id=71 lang=javascript
 *
 * [71] Simplify Path
 */

// @lc code=start
/**
 * @param {string} path
 * @return {string}
 */
var simplifyPath = function (path) {
  const stack = [];
  const tokens = path.split("/");
  for (const token of tokens) {
    if (token === "..") {
      if (stack.length) {
        stack.pop();
      }
    } else if (token && token !== ".") {
      stack.push(token);
    }
  }
  return "/" + stack.join("/");
};
// @lc code=end
